SCAutolib/models/CA.py

```diff
@@ -196,3 +207,322 @@ def cleanup(self):
+                            "previous installation.")
+                return
+            self.restore()
+
+        logger.info(f"Start setup of IPA client on the system for "
+                    f"{self._ipa_server_hostname} IPA server.")
+
+        self._add_to_resolv()
+        self._set_hostname()
+
+        run(["ipa-client-install", "-p", "admin",
+             "--password", self._ipa_server_admin_passwd,
+             "--server", self._ipa_server_hostname,
+             "--domain", self._ipa_server_domain,
+             "--realm", self._ipa_server_realm,
+             "--hostname", self._ipa_client_hostname,
+             "--all-ip-addresses", "--force", "--force-join", "--no-ntp", "-U"],
+            input="yes")
+        logger.debug("IPA client is installed")
+
+        ipa_client_script = self._get_sc_setup_script()
+        run(f'bash {ipa_client_script} /etc/ipa/ca.crt')
+        logger.debug("Setup of IPA client for smart card is finished")
+
+        policy = self.meta_client.pwpolicy_show(a_cn="global_policy")["result"]
+        if "0" not in policy["krbminpwdlife"]:
+            self.meta_client.pwpolicy_mod(a_cn="global_policy",
+                                          o_krbminpwdlife=0)
+            logger.debug("Minimal kerberos password lifetime is set to 0 days")
+        if "365" not in policy["krbmaxpwdlife"]:
+            self.meta_client.pwpolicy_mod(a_cn="global_policy",
+                                          o_krbmaxpwdlife=365)
+            logger.debug("Maximum kerberos password lifetime is set to 365 days")
+
+        # TODO: add to restore client host name
+        logger.info("IPA client is configured on the system.")
+
+    @property
+    def is_installed(self):
+        """
+        :return: True, if IPA client is installed on the system (ipa command
+            returns zero return code), otherwise False
+        :rtype: bool
+        """
+        out = run(["ipa", "help"], print_=False)
+        return out.returncode == 0
+
+    def _set_hostname(self):
+        """
+        Set hostname for specified IPA client hostname
+        """
+        run(f"hostnamectl set-hostname {self._ipa_client_hostname} --static")
+        logger.debug(f"Hostname is set to {self._ipa_client_hostname}")
+
+    def _add_to_resolv(self):
+        """
+        Add new nameserver (IPA) to /etc/resolv.conf and lock this file for
+        editing
+        """
+        nameserver = f"nameserver {self._ipa_server_ip}"
+        with open("/etc/resolv.conf", "w+") as f:
+            cnt = f.read()
+            if nameserver not in cnt:
+                logger.warning(f"Nameserver {self._ipa_server_ip} is not "
+                               "present in /etc/resolve.conf. Adding...")
+                f.write(nameserver + "\n" + cnt)
+                logger.info(
+                    "IPA server is added to /etc/resolv.conf "
+                    "as first nameserver")
+                run("chattr -i /etc/resolv.conf")
```

Comment:
what's the cleanup/reexecutability story here?
why is a `chattr -i` needed?

Reply:
There was a problem with the IPA server that the client host couldn't resolve the hostname of the IPA server because NetworkManager rewrite the resolv.conf after it has be updated by IPA script. So, this piece of code aims to add IPA server to resolvers and then look this file from editing by `chattr -i...`. Do you have a better solution for this problem?